# Sqlite3

o Sqlite3 é uma versão *light* de SQL que vem integrada como uma biblioteca standard de Python. SQL é uma linguagem de programação muito util para trabalharmos com bases de dados.

Uma base de dados não é nada mais que um conjunto de tabelas com valores.

Para utilizar começamos por importar a biblioteca:

In [5]:
import sqlite3

def read_file(file):
    player_list = []
    
    f = open(file, "r")
    lines = f.readlines()
    
    for line in lines:
        aux_list = line.split(",")
        
        aux_dict = {}
        aux_dict["name"] = aux_list[0]
        aux_dict["age"] = int(aux_list[1])
        aux_dict["total_min"] = int(aux_list[7])
        aux_dict["nacionality"] = aux_list[10]
        aux_dict["total_games"] = int(aux_list[11])
        aux_dict["total_goals"] = int(aux_list[14])
        
        player_list.append(aux_dict)
    f.close()
    return player_list

player_list = read_file("EPL1819.csv")

Criar uma conecção e definir um cursor:

In [7]:
conn = sqlite3.connect("players.db") # Criar a ligação com uma base de dados
c = conn.cursor() # Criar o cursor

Quando criamos uma **conecção** com uma base de dados, **se a base de dados não existir esta é criada automaticamente**.

Um **cursor** é equivalente ao "cursor do rato", dado que é aquilo que nos **vai permitir fazer diferentes acções**.

Criar uma tabela:

In [21]:
c.execute("CREATE TABLE IF NOT EXISTS Players(name TEXT, age INTEGER, total_min INTEGER, nacionality TEXT,\
total_games INTEGER, total_goals INTEGER)")

Tipos de dados em SQL:

    - INTEGER: número inteiro;
    - REAL: número real;
    - TEXT: texto.

Adicionar entradas numa tabela - **INSERT INTO**:

In [25]:
#c.execute("INSERT INTO TestTable VALUES(101,'2020-05-29 13:53:39','Python',6.1)")

*.commit()* é equivalente a guardarmos as alterações que fizemos.

Introduzir variáveis numa tabela:

In [22]:
for player in player_list:
    
    c.execute("INSERT INTO Players (name, age, total_min, nacionality, total_games, total_goals)\
    VALUES (?, ?, ?, ?, ?, ?)",
              (player["name"], player["age"], player["total_min"], player["nacionality"], player["total_games"], player["total_goals"]))

In [23]:
conn.commit()

Ler conteúdo de uma tabela - **SELECT**:

In [24]:
c.execute('SELECT * FROM Players')
data = c.fetchall()

print(data)

[('David de Gea', 29, 3420, 'Spain', 38, 0), ('Matteo Darmian', 30, 443, 'Italy', 6, 0), ('Victor Nilsson Lindelöf', 25, 2602, 'Sweden', 30, 1), ('Luke Shaw', 24, 2592, 'England', 29, 1), ('Eric Bertrand Bailly', 26, 637, "Côte d'Ivoire", 12, 0), ('Juan Mata', 31, 1294, 'Spain', 22, 3), ('Paul Pogba', 27, 3011, 'France', 35, 13), ('Frederico Rodrigues Santos', 27, 1045, 'Brazil', 17, 1), ('Andreas Pereira', 24, 595, 'Brazil', 15, 1), ('Alexis Sanchez', 31, 877, 'Chile', 20, 1), ('Marcus Rashford', 22, 2362, 'England', 33, 10), ('Romelu Lukaku', 26, 2130, 'Belgium', 32, 12), ('Scott McTominay', 23, 861, 'England', 16, 2), ('Marouane Fellaini', 32, 744, 'Belgium', 14, 0), ('Anthony Martial', 24, 1623, 'France', 27, 10), ('Chris Smalling', 30, 2128, 'England', 24, 1), ('Lee Grant', 37, 0, 'England', 0, 0), ('Ashley Young', 34, 2568, 'England', 30, 2), ('Kasper Schmeichel', 33, 3420, 'Denmark', 38, 0), ('Wes Morgan', 36, 1926, 'Jamaica', 22, 3), ('Harry Maguire', 27, 2599, 'England', 31, 3

In [25]:
for row in data:
    print(row)

('David de Gea', 29, 3420, 'Spain', 38, 0)
('Matteo Darmian', 30, 443, 'Italy', 6, 0)
('Victor Nilsson Lindelöf', 25, 2602, 'Sweden', 30, 1)
('Luke Shaw', 24, 2592, 'England', 29, 1)
('Eric Bertrand Bailly', 26, 637, "Côte d'Ivoire", 12, 0)
('Juan Mata', 31, 1294, 'Spain', 22, 3)
('Paul Pogba', 27, 3011, 'France', 35, 13)
('Frederico Rodrigues Santos', 27, 1045, 'Brazil', 17, 1)
('Andreas Pereira', 24, 595, 'Brazil', 15, 1)
('Alexis Sanchez', 31, 877, 'Chile', 20, 1)
('Marcus Rashford', 22, 2362, 'England', 33, 10)
('Romelu Lukaku', 26, 2130, 'Belgium', 32, 12)
('Scott McTominay', 23, 861, 'England', 16, 2)
('Marouane Fellaini', 32, 744, 'Belgium', 14, 0)
('Anthony Martial', 24, 1623, 'France', 27, 10)
('Chris Smalling', 30, 2128, 'England', 24, 1)
('Lee Grant', 37, 0, 'England', 0, 0)
('Ashley Young', 34, 2568, 'England', 30, 2)
('Kasper Schmeichel', 33, 3420, 'Denmark', 38, 0)
('Wes Morgan', 36, 1926, 'Jamaica', 22, 3)
('Harry Maguire', 27, 2599, 'England', 31, 3)
('Ricardo Domingos 

O operador **SELECT** é seguido das colunas da tabela que queremos retornar, por exemplo:

    - * devolve todas as colunas;
    - name devolve apenas essa coluna;
    - name, total_goals devolve a coluna id e a coluna words.

In [27]:
c.execute('SELECT total_goals, name FROM Players') # A ordem com que colocamos as colunas tem inflencia no resultado
data = c.fetchall()

print(data)

[(0, 'David de Gea'), (0, 'Matteo Darmian'), (1, 'Victor Nilsson Lindelöf'), (1, 'Luke Shaw'), (0, 'Eric Bertrand Bailly'), (3, 'Juan Mata'), (13, 'Paul Pogba'), (1, 'Frederico Rodrigues Santos'), (1, 'Andreas Pereira'), (1, 'Alexis Sanchez'), (10, 'Marcus Rashford'), (12, 'Romelu Lukaku'), (2, 'Scott McTominay'), (0, 'Marouane Fellaini'), (10, 'Anthony Martial'), (1, 'Chris Smalling'), (0, 'Lee Grant'), (2, 'Ashley Young'), (0, 'Kasper Schmeichel'), (3, 'Wes Morgan'), (3, 'Harry Maguire'), (2, 'Ricardo Domingos Barbosa Pereira'), (0, 'Ben Chilwell'), (0, 'Adrien Sebastian Perruchet Silva'), (0, 'Daniel Amartey'), (2, 'Onyinye Wilfred Ndidi'), (4, 'Demarai Gray'), (7, 'James Maddison'), (1, 'Kelechi Iheanacho'), (18, 'Jamie Vardy'), (1, 'Rachid Ghezzal'), (0, 'Vicente Iborra De La Fuente'), (1, 'Jonny Evans'), (2, 'Marc Albrighton'), (0, 'Danny Ward'), (0, 'Christian Fuchs'), (0, 'Martin Dúbravka'), (3, 'Ciaran Clark'), (0, 'Jamal Lascelles'), (0, 'Paul Dummett'), (1, 'DeAndre Yedlin')

Temos de utilizar sempre o *.fetchall()* após a pesquisa para recebermos a lista dos resutados.

Podemos adicionar alguma lógica às nossas pesquisas utilizando o operador **WHERE**:

In [31]:
c.execute('SELECT * FROM Players WHERE total_goals > 20')
data = c.fetchall()

print(data)

[('Mohamed Salah', 27, 3262, 'Egypt', 38, 22), ('Sadio Mané', 28, 3086, 'Senegal', 36, 22), ('Pierre-Emerick Aubameyang', 30, 2731, 'Gabon', 36, 22), ('Sergio Aguero', 31, 2480, 'Argentina', 33, 21)]


Podemos utilizar o operador **LIKE** para pesquisar termos semelhantes a um determidado valor:

In [34]:
c.execute('SELECT * FROM Players WHERE name LIKE "%David%" ORDER BY total_goals;')
data = c.fetchall()

print(data)

[('David de Gea', 29, 3420, 'Spain', 38, 0), ('Davide Zappacosta', 27, 125, 'Italy', 4, 0), ('David Button', 31, 360, 'England', 4, 0), ('Adam David Lallana', 31, 465, 'England', 13, 0), ('David Junior Hoilett', 29, 2142, 'Canada', 32, 3), ('David Luiz', 33, 3239, 'Brazil', 36, 3), ('David Silva', 34, 2411, 'Spain', 33, 6), ('David Brooks', 22, 2278, 'England', 30, 7)]


In [37]:
c.execute('SELECT DISTINCT nacionality FROM Players')
data = c.fetchall()

print(data)

[('Spain',), ('Italy',), ('Sweden',), ('England',), ("Côte d'Ivoire",), ('France',), ('Brazil',), ('Chile',), ('Belgium',), ('Denmark',), ('Jamaica',), ('Portugal',), ('Ghana',), ('Nigeria',), ('Algeria',), ('Northern Ireland',), ('Wales',), ('Austria',), ('Slovakia',), ('Republic of Ireland',), ('USA',), ('Senegal',), ('Scotland',), ('Venezuela',), ('Japan',), ('South Korea',), ('Switzerland',), ('Colombia',), ('Netherlands',), ('Bosnia and Herzegovina',), ('Norway',), ('Poland',), ('Philippines',), ('Gabon',), ('Canada',), ('Germany',), ('Serbia',), ('Argentina',), ('Congo DR',), ('Australia',), ('Benin',), ('Egypt',), ('Slovenia',), ('Greece',), ('Cameroon',), ('Mali',), ('Iran',), ('Israel',), ('Mexico',), ('Morocco',), ('Iceland',), ('Turkey',), ('Guinea',), ('Paraguay',), ('Ukraine',), ('Equatorial Guinea',), ('New Zealand',), ('Czech Republic',), ('Armenia',), ('Uruguay',), ('Croatia',), ('Montenegro',), ('Ecuador',), ('Kenya',), ('Togo',), ('Romania',), ('Curaçao',)]


In [42]:
c.execute('SELECT name, max(total_games) FROM Players WHERE total_goals>20;')
data = c.fetchall()

print(data)

[('Mohamed Salah', 38)]


Ver páginas 125 a 127 das folhas da disciplina.

No final de termos as operações finalizadas devemos fazer *.commit()* e fechar o cursos e a connection.

In [43]:
conn.commit()
c.close()
conn.close()